### ISSUE TITLE
Expose watch api in the js library

### ISSUE THREAD

**Issue by akkumar** [sass/sass#2823]
Right now the js library exposes render and renderSync api. 

it would be useful to have a "watch" api  exposed as well.

There are a bunch of api available in `watch.dart` but not in a way directly consumable by the js api. 

It would be useful to expose the same.

**Comment by nex3** [comment]
I've moved this to the Sass language repo because the JS API is shared across multiple Sass implementations to ensure cross-compatibility.

I feel like adding a public API for filesystem watching is outside the domain of a Sass implementation, which should primarily focus on the business of compiler logic. There are many watcher packages that already exist in JavaScript, including both stand-alone libraries like Chokidar (which Sass uses) and build systems like Webpack that have plugins that allow them to work with Sass. What's the benefit of re-exposing this functionality from within Sass rather than allowing users to use Sass's compilation API in conjunction with whatever watcher API they choose? I don't see it.

@saper, do you have a strong opinion here as node-sass maintainer?

**Comment by xzyfer** [comment]
We've previously discussed this within node-sass. Our plan, keeping in line with existing Node CLIs, has been to expose a `watchConfig` object on the render config object that we would blindly pass along to chokidar. 

We've been blocked waiting for the migration from gaze to chokidar to land. It has since landed on upcoming v5 release branch.

**Comment by nex3** [comment]
What's the benefit of just passing configuration straight to Chokidar rather than having users configure Chokidar on their own?

**Comment by xzyfer** [comment]
We've typically treated node-sass as developer CLI as opposed to a bare Sass compiler like LibSass. As a result we've baked in developer productivity features like watching and incremental compilation.

We use some watcher APIs to achieve incremental compilation so we prefer decorate watcher configuration than mutate or decorate a watcher instance.

**Comment by nex3** [comment]
> We've typically treated node-sass as developer CLI as opposed to a bare Sass compiler like LibSass. As a result we've baked in developer productivity features like watching and incremental compilation.

What APIs does Node Sass expose for incremental compilation? I don't see anything about this in the README.

> We use some watcher APIs to achieve incremental compilation so we prefer decorate watcher configuration than mutate or decorate a watcher instance.

WDYT about making a separate package like `chokidar-sass` that exposes these APIs in a cross-implementation way, rather than requiring each implementation to re-implement them?

**Comment by xzyfer** [comment]
>What APIs does Node Sass expose for incremental compilation?

We don't expose any APIs. It happens by default when watching. To achieve this we integrate with the watcher APIs. This is partly responsible for the delayed migration from Gaze to Chokidar.

>WDYT about making a separate package like

We're open to it. We've discussed doing so in node-sass for a while. Even so we'd like have single node-sass CLI for watching and compilation like we have today. 

That's why our preference has been to accept a watchConfig object, so that on the CLI we could also have `--watch-<chokidar-option>=<chokidar-value>`.

**Comment by saper** [comment]
Thanks @nex3 for bringing this up. Indeed there are some cross-platform issues we should somehow address. 

I've been collecting all node-sass issues related to dependency management/watching under the [dependency tree management](https://github.com/sass/node-sass/projects/3) project.

What is missing is a specification of a Sass build system or a project (possibly object) model.

I could imagine such a system to provide unified semantics to determine the following:

1. Where does the source bytestream come from (URLs, files, programmatically generated - custom importers)
1. Whether we deal with SASS or SCSS syntax (currently mostly guessed based on a filename extension, but see 
1. Whether a given module/file is a "main" file or includeable entity that will not compile on its own (partial, ...), otherwise we suffer from a major node-sass watcher bug that is https://github.com/sass/node-sass/issues/2491
1. How to feed the compiler with a source stream (push model, pull model)
1. How does dependency tree look like
1. Which parts can be compiled safely in parallel

Watching is basically just a bidirectional interface that 
1. feeds dependency tree changes to some external entity
1. gets feedback from the entity to kick compilation of some modules

This way we should get rid of things like sass-graph, we could get a clean importer interface, etc. etc.

What could be cross-platform?
- project definition ("those are main SCSS files", "those are includeble SASS libraries", "here is a program to import otherwise unknown modules")
- maybe we arrive at a common file format (something like `pyproject.toml`, maybe)
- we should have a common interface that compilers expose to the build system ("what are the dependencies for X?", "compile Y as SASS/SCSS".
- maybe (out of scope for this bug) we could even arrive at some common abstract syntax tree interface

The build system could, in turn, interface with watchers and other interested parties.

**Comment by nex3** [comment]
@xzyfer 

> We don't expose any APIs. It happens by default when watching. To achieve this we integrate with the watcher APIs. This is partly responsible for the delayed migration from Gaze to Chokidar.\

Ah, gotcha.

> We're open to it. We've discussed doing so in node-sass for a while. Even so we'd like have single node-sass CLI for watching and compilation like we have today.
> 
> That's why our preference has been to accept a watchConfig object, so that on the CLI we could also have `--watch-<chokidar-option>=<chokidar-value>`.

Okay, cool. I think it's fine for each implementation to expose their own CLI options--it's just the JS API that I think should be coordinated.

@saper All of those are good questions. I think some of them will be addressed by #2509, and are probably already partially addressed by the more explicit specs around [loading source files](https://github.com/sass/sass/blob/master/spec/modules.md#loading-a-source-file), although those specs currently don't address importers at all.

I think I'd be happy ending up somewhere where we have a unified API for a `Compiler` object (as outlined in #2745) that also has a push-based API for indicating "this file has changed". That would limit the API surface to something pretty self-contained while still providing enough flexibility to implement a thin watcher wrapper on top in a separate `sass-chokidar` package.

**Comment by akkumar** [comment]
So - any consensus on this and any time frame for this API to land in this package (or a separate package )  ? 

It would be useful for a lot of downstream projects  / IDE plugins to have this.

**Comment by saper** [comment]
@akkumar feel free to build a separate package that does this

**Comment by nex3** [comment]
@akkumar Designing and implementing new APIs takes a lot of work, and the Sass team has limited bandwidth. If you want to drive this feature, though, feel free to write up a proposal!

**Comment by nex3** [comment]
Merging this into https://github.com/sass/sass/issues/2745.
